---
title: "Wrens WGS: Descriptive Statistics"
output: html_notebook
---


<center>

<h1>Exploratory Analysis WGS of Wrens</h1>

</center>

<center>

<h2>Luis Daniel Montalvo</h2>

</center>


### Introduction

This is the general analysis of the WGS data for Campylorhynchus zoantus and fasciatus. Sequencing was done in ICBR at the university of Florida. Protocols for the data processing are explained in the document Notes on Molecular Analyses - wgs.docx. Here we analyze the resulting vcf file obtained with this protocol.


```{r setup, include=FALSE}

rm(list=ls())

setwd("C:/Users/Daniel/Dropbox/Thesis/Molecular_Wrens/WGS/R/LEA")

load("WGS_LEA.RData")

```



```{r General Data Management, include=FALSE}


install.packages("raster")
library(raster)

##################
## Making POP STRUCTURE file for Tassel

## Read samples names equivalents to the radseq data.
wgs.qvalues <- read.csv("../Predictors/pop.struc.csv", sep=",", header=TRUE)

## Writting the population structure files for TASSEL
write.csv(wgs.qvalues[,c(1:2,9:13)],  file="wgs.qvalues.csv")


##################
## Getting Climate Data

## Getting name of groups and coodinates
gr.coor <- wgs.qvalues[,c("Ind", "Lat", "Long")]

colnames(gr.coor) <- c("Ind", "lon", "lat")

## Making the coordinates numeric
coor <- cbind(as.numeric(as.character(gr.coor$lon)), as.numeric(as.character(gr.coor$lat)))

## Setting the coordinates system as Long/Lat
cord.dec = SpatialPoints(cbind(coor[,2], coor[,1]), proj4string=CRS("+proj=longlat +datum=WGS84 +no_defs +ellps=WGS84"))


## Downloading CHELSA data
Chelsa.Clim.download(parameter = "bio",
                     bio.var =  c(1:19), 
                     version.var = "1.2", 
                     clipping = TRUE, 
                     clip.extent = c(-112, -79, -7, 35), 
                     buffer = 0, 
                     convert.files.to.asc = FALSE, 
                     stacking.data = TRUE, 
                     combine.raw.zip = FALSE,
                     delete.raw.data = FALSE,
                     save.bib.file = TRUE)


## Setting the directory where the raster was saved
wd <- ("C:\\Users\\Daniel\\Dropbox\\Thesis\\Molecular_Wrens\\WGS\\R\\bio\\bio_V1.2\\")

## Creating a list with the names of all raster files 
list.raster <- list.files(wd, full.names = TRUE)

## Reading and stacking the rasters
stack.ch <- stack(list.raster[1:19])

## Extracting the values for the sampling points I have
values.ch <- raster::extract(stack.ch, cord.dec)
View(values.ch)

## Naming the columns
colnames(values.ch) <- c("AMT","MDR","ISO","TS","MTWM","MTCM","TAR","MTWetQ","MTDQ","MTWarQ","MTCQ","AMP","PWetM","PDM","PS","PWetQ","PDQ","PWarQ","PCQ")

## Merging the coordinates, Chelsa values, population labels
coor.clim.raw.ch <- cbind.data.frame(coordinates(coor), values.ch)
colnames(coor.clim.raw.ch)[1:2] <- c("Long", "Lat")
coor.clim.raw.ch <- cbind(gr.coor, coor.clim.raw.ch)
clim <- merge(coor.clim.raw.ch, wgs.qvalues, by="Ind")
clim <- clim[,c("Ind", "sampleMuseum", "lat", "AMT", "AMP", "PS")]

## Writting the climate data for TASSEL
write.csv(clim,  file="clim.chelsa.csv")

install.packages("PerformanceAnalytics")
library(PerformanceAnalytics)

# Plotting multiple correlations
png("corr_19clim.jpg", width = 10000, height = 10000, units = "px", res=300)
chart.Correlation(coor.clim.raw.ch[,6:24], method = "pearson", histogram = TRUE, pch = 16)
dev.off()

#chart.Correlation(clim[,c( "AMT", "AMP", "PS")], method = "pearson", histogram = TRUE, pch = 16)

##################
## Climate of Worldclim
r <- getData("worldclim", var="bio", res=10)
r3 <- r[[c(1,12,15)]]
names(r3) <- c("AMT", "AMP", "PS")
colnames(cord.dec@coords) <- c("Y", "")
wgs.wclim <- raster::extract(r3, cord.dec@coords[,2:1])
wgs.wclim <- cbind.data.frame(gr.coor, wgs.wclim)
wgs.wclim <- merge(wgs.wclim, wgs.qvalues, by="Ind")
wgs.wclim <- wgs.wclim[,c("Ind", "sampleMuseum", "lat", "AMT", "AMP", "PS")]

## Writting the climate data for TASSEL
write.csv(wgs.wclim,  file="wgs.wclim.csv")

## The coding as it will go in Tassel is this
## CZ	CFPn	CFPs	CFF	CB
## Q1	Q2	Q3	Q4	Q5


```



############################    LEA ANALYSIS   ##############################

#### GWAS with LEA

```{r GWAS with LEA: testing and Data Imputation, include=FALSE}

install.packages("pegas")
library(pegas)

## Read vcf file
wrens.vcf <- read.vcf(path2vcf)

## Installing LEA package
if (!require("BiocManager", quietly = TRUE))
    install.packages("BiocManager")

BiocManager::install("LEA")

library(LEA)

## Convert vcf lfmm object
path2vcf <- "C:/Users/Daniel/Dropbox/Thesis/Molecular_Wrens/WGS/R/Genotype/wgs.wrens.named.vcf"
vcf2lfmm(path2vcf)

## Read table txt file
wrens.env <- read.table("C:/Users/Daniel/Dropbox/Thesis/Molecular_Wrens/WGS/R/LEA/clim3var.97s.env", header = TRUE)

## REad table with order of sample
sample.ord <- read.table("C:/Users/Daniel/Dropbox/Thesis/Molecular_Wrens/WGS/R/LEA/sampleOrd.txt", header = TRUE)

## Order wrens.env according to sample.ord$Sample
## It has to have the same order as vcf file
wrens.env <- wrens.env[match(sample.ord$Sample, wrens.env$Sample),]

## Check if the order is correct
dim(sample.ord)
head(wrens.env)

## Write the env file
LEA::write.env(wrens.env[,2:4], "C:/Users/Daniel/Dropbox/Thesis/Molecular_Wrens/WGS/R/LEA/chelsa3var97samples.env")

##############################################################################################################
## Testing the LFMM

## Read lfmm and climate data
patch.lfmmm.chr7 <- "C:/Users/Daniel/Dropbox/Thesis/Molecular_Wrens/WGS/R/LEA/wgs.wrens.chr7.lfmm"
patch.env <- "C:/Users/Daniel/Dropbox/Thesis/Molecular_Wrens/WGS/R/LEA/chelsa3var97samples.env"

## Running the LFMM
ml.chr7.test <- lfmm2(input = patch.lfmmm.chr7, 
                env = patch.env, 
                K = 2)

## Getting pvalues
pv.chr7 <- lfmm2.test(object = ml.chr7.test, input = patch.lfmmm.chr7, env = patch.env, 
                 full = TRUE, linear = TRUE)

## Plotting the test
plot(-log10(pv.chr7$pvalues), col = "grey", cex = .6, pch = 19)
abline(h = -log10(0.1/510), lty = 2, col = "orange")

##############################################################################################################
## The LFMM can't handle missing data, so we need to impute the missing data

## DATA IMPUTATION

## Entire loci data
patch.lfmm <- "C:/Users/Daniel/Dropbox/Thesis/Molecular_Wrens/WGS/R/LEA/wgs.wrens.named.lfmm"
patch.geno <- "C:/Users/Daniel/Dropbox/Thesis/Molecular_Wrens/WGS/R/LEA/wgs.wrens.named.geno"

imp.snmf <- snmf(patch.lfmm,
                 K=4,
                 entropy = TRUE,
                 repetitions = 10,
                 project = "new",
                 CPU = 5)

## Select the run with the lowest cross-entropy value
best.snmf = which.min(cross.entropy(imp.snmf, K = 4))


## Impute the missing genotypes
impute(imp.snmf, patch.lfmm, method = 'mode', K = 4, run = best.snmf)

```


#### GWAS with LEA: Model with full = TRUE


```{r GWAS with LEA: Model with full = TRUE, include=FALSE}


##############################################################################################################
## Running the LFMM with the imputed data

## Entire loci data
patch.lfmm.imp <- "C:/Users/Daniel/Dropbox/Thesis/Molecular_Wrens/WGS/R/LEA/wgs.wrens.named.lfmm_imputed.lfmm"

## Run LFMM
ml.clim.1 <- lfmm2(input = patch.lfmm.imp, 
                env = patch.env, 
                K = 4)

#pv.amt <- lfmm.pvalues(ml.amt.1, K = 4)
#pvalue.amt1 <- length(pv.amt$pvalues)

#par(mfrow = c(2,1))
#hist(pv.amt$pvalues, col = "lightblue")
#plot(-log10(pv.amt$pvalues), pch = 19, col = "blue", cex = .7)

## Get p-values
pv.env1 <- lfmm2.test(object = ml.clim.1, input = patch.lfmm.imp, env = patch.env, 
                 full = TRUE, linear =  TRUE)

## We get NAs in the pvalues beacuse of the invariant sites that produced NA in adj.r.squared

## Get the type of variables in pv.env1
str(pv.env1)

## Check the issue with NA
sum(is.na(pv.env1$pvalues))
length(pv.env1$pvalues)

##############################################################################################################

## I see that there are seven NA in adj.r.squared
## I will remove them from the analysis pml.clim.1
## First I have to find the index  in pv.env1$add.r.squared taht are NA

## Get the index of the adj_ that are NA
index.na <- which(is.na(pv.env1$adj.r.squared))

## Read table lfmm
lfmm.full.ori <- read.table("C:/Users/Daniel/Dropbox/Thesis/Molecular_Wrens/WGS/R/LEA/wgs.wrens.named.lfmm_imputed.lfmm", header = FALSE)

## Checking dimensions
dim(lfmm.full.ori)

## Remove the rows with NA that are invariant sites
lfmm.invariant <- lfmm.full.ori[,-index.na]

## Checking dimensions
dim(lfmm.invariant)

## Write the new table
write.table(lfmm.invariant, "C:/Users/Daniel/Dropbox/Thesis/Molecular_Wrens/WGS/R/LEA/wgs.wrens.named.imputed.invariant.lfmm", row.names = FALSE, col.names = FALSE)

## Entire loci data
patch.lfmm.invariant <- "C:/Users/Daniel/Dropbox/Thesis/Molecular_Wrens/WGS/R/LEA/wgs.wrens.named.imputed.invariant.lfmm"

##############################################################################################################

## Run LFMM Again without invariant sites
ml.clim.2 <- lfmm2(input = patch.lfmm.invariant, 
                env = patch.env, 
                K = 4)

## Get p-values
pv.env2 <- lfmm2.test(object = ml.clim.2, input = patch.lfmm.invariant, env = patch.env, 
                full = TRUE, linear =  TRUE)

## Get the type of variables in pv.env1
str(pv.env2)

## We finally got the p-values

## Plotting p-values
png("Manhattan_env2.jpg", width = 10000, height = 6000, units = "px", res=300)
par(mfrow=c(1,1), mar=c(8,8,5,5))
LEA::plot(-log10(pv.env2$pvalues), col = "grey", cex = .6, pch = 19,
          xlab = "SNPs", ylab = "-log10(p-value)", main = "Manhattan plot of p-values") 
abline(h = -log10(0.05/length(pv.env2$pvalues)), lty = 2, col = "orange")
dev.off()

## Making qqplot
qqplot(rexp(length(pv.env2$pvalues), rate = log(10)),
       -log10(pv.env2$pvalues), xlab = "Expected quantile",
       pch = 19, cex = .4)
abline(0,1)


##############################################################################################################
## We want to get the names and positions of the SNPs to merge it with the p-values

## We will use the vcfR package
library(vcfR)

## Read vcf file
vcf <- read.vcfR("C:/Users/Daniel/Dropbox/Thesis/Molecular_Wrens/WGS/R/Genotype/wgs.wrens.named.vcf")

vcf

## Get dataframe with chromosomes names and positions and markers names
vcf.df <- data.frame(vcf@fix[,c(1,2,3)])
head(vcf.df)

## Some exploratory plots for missing data
install.packages("SNPfiltR")
library(SNPfiltR)

## Get missing data per sample and site
miss.per.sam <- missing_by_sample(vcf)
missing_by_snp(vcf)

dim(miss)
View(miss)

## Read the files we got from the unix commands
## vcftools --vcf wgs.wrens.named.vcf  --missing-indv
## vcftools --vcf wgs.wrens.named.vcf  --missing-site

## Read table outimiss
outimiss <- read.table("out.imiss", header = TRUE)
View(outimiss)

## Get average of percentage of missing data per sample
mean(outimiss$F_MISS)

## REad table outlmiss
outlmiss <- read.table("out.lmiss", header = TRUE)
View(outlmiss)

## Get average of percentage of missing data per site
mean(outlmiss$F_MISS)
dim(outlmiss)
head(outlmiss)

## Read plink.counts file with heterozygosity per sample from vcf
plink.counts <- read.table("plink2.scount", header = TRUE)
View(plink.counts)

## Add collumn with percentage of heterozygosity
plink.counts$HET <- plink.counts$HET_CT/(plink.counts$HOM_CT+plink.counts$HET_CT)

## Get mean of Het
mean(plink.counts$HET)

##############################################################################################################

## Read the file with Snps removed when we converted the vcf above with vcf2lfmm
snps.removed <- read.table("C:/Users/Daniel/Dropbox/Thesis/Molecular_Wrens/WGS/R/LEA/wgs.wrens.named.removed", header = FALSE)
head(snps.removed)

## Remove row in vcf.df where column ID are in snps.removed
vcf.df.2 <- vcf.df[!(vcf.df$ID %in% snps.removed$V3),]
head(vcf.df.2)

## Checking dimensions
dim(vcf.df)
dim(vcf.df.2)

## Checking if the number of rows in vcf.df.2 is equal to the number of rows in snps.removed
identical(length(vcf.df$ID), length(vcf.df.2$ID)+length(snps.removed$V3))

## Remove the rows in vcf.df.2 that has the index index.na
vcf.df.3 <- vcf.df.2[-index.na,]
head(vcf.df.3)
dim(vcf.df.3)

## Putting together the chromosomes names in vcf.df.3 and the pvalues and adj.r.squared from pv.env2
chr.pval.env2 <- data.frame(vcf.df.3, pv.env2$pvalues, pv.env2$adj.r.squared)
dim(chr.pval.env2)
head(chr.pval.env2)

## Change the names of the columns
names(chr.pval.env2) <- c("CHROM", "POS", "ID", "pvalue", "adj.r.squared")

## Order the dataframe by pvalue from higher to lower
chr.pval.env2.ordered <- chr.pval.env2[order(chr.pval.env2$pvalue),]
head(chr.pval.env2.ordered)


```


#### GWAS with LEA: Model with full = FALSE

Get pvalues adjusted for multiple testing

Now we want the pvalues per climate variable to see what SNPs are associated to each climate variable. 
We make this by setting the full = FALSE in the lfmm2.test function.

Use Benjamini & Hochberg (1995) procedure to estimate the adjusted p-values
The procedure is based on the assumption that the tests are independent or positively dependent.
The procedure controls the FDR at level q, and guarantees that the expected proportion of 
false discoveries among the rejected hypotheses does not exceed q

An explanation how to estimate the Benjamini & Hochberg (1995) procedure is here:
https://www.statology.org/benjamini-hochberg-procedure/


```{r GWAS with LEA: Per Cliamte Variable, include=FALSE}

## Get p-values per climate variable
library(LEA)
pv.env3.gc <- lfmm2.test(object = ml.clim.2, input = patch.lfmm.invariant, env = patch.env, 
                full = FALSE, linear =  TRUE, genomic.control = TRUE)

str(pv.env3.gc)

str(pv.env3.gc$pvalues)
class(pv.env3.gc$pvalues)

## Check structure of pv.env3$pvalues three rows and nloci columns, here just 10 columns
pv.env3.noGC$pvalues[,1:10]
## The three first values are the pvalues of the three climate variables
## Make data frame byrow=TRUE to put these three values in the same row

## Putting together chr.pval.env2 and the pvalues and zscores in pv.env3
chr.pval.env3 <- data.frame(chr.pval.env2, 
                            matrix(pv.env3.gc$pvalues, ncol = 3, byrow = TRUE),
                            matrix(pv.env3.gc$zscores, ncol = 3, byrow = TRUE))

head(chr.pval.env3)

## Change the names of the columns
names(chr.pval.env3)[6:11] <- c("pval_AMT", "pval_AMP", "pval_PS", "zscore_AMT", "zscore_AMP", "zscore_PS")

## The histogram indicate that the pvalues are correctly adjusted 
## Pvalues show a peack at 0.05 and then decrease
par(mfrow=c(2,2), mar=c(8,8,5,5))
hist(chr.pval.env3$pvalue)
hist(chr.pval.env3$pval_AMT)
hist(chr.pval.env3$pval_AMP)
hist(chr.pval.env3$pval_PS)

```



#### AMT

```{r GWAS with LEA: AMT, include=FALSE}

##############################################################################################################
## Annual Mean Temperature (AMT)

## FDR of 0.05
q = 0.05

## Order sig.snps.amt by pval_AMT
pval.env3.ordered.by.amt <- chr.pval.env3[order(chr.pval.env3$pval_AMT),]

## Number of SNPs
L.amt = length(pval.env3.ordered.by.amt$pval_AMT)

## Adding the FDR adjusted pvalue
pval.env3.ordered.by.amt$BHadj.amt <- q*(1:L.amt)/L.amt
head(pval.env3.ordered.by.amt)

str( pval.env3.ordered.by.amt)
pval.env3.ordered.by.amt$CHROMF <- as.numeric(as.factor(pval.env3.ordered.by.amt$CHROM))
pval.env3.ordered.by.amt$POS <- as.numeric(pval.env3.ordered.by.amt$POS)

## Indices for Candidates SNPs
w.amt = which(pval.env3.ordered.by.amt$pval_AMT < q*(1:L.amt)/L.amt)
length(w.amt)

## Get rows in pval.env3.ordered.by.amt that correspond to the candidate SNPs
candidates.amt <- pval.env3.ordered.by.amt[w.amt,]
View(candidates.amt)


########################
## Manhattan and QQPlots

library(qqman)


## Labels for chromosomes names in Manhattan plots
chr_labels <- unique(chr.pval.env3$CHROM)

## AMT
tail(candidates.amt)

png("manhattan_AMT2.jpg", width = 12000, height = 6000, units = "px", res=300)
par(mfrow=c(1,1), mar=c(8,8,5,5))
manhattan(pval.env3.ordered.by.amt, chr="CHROMF", bp="POS", snp="ID", 
                        genomewideline = F, suggestiveline = F,
                        logp = T, p="pval_AMT", annotateTop = T, main = "Manhattan Plot - AMT", 
                        ylim=c(0,10), cex.lab=3, cex.axis=2, cex.main=3, cex.sub=3,
                        col = c("blue4", "goldenrod1"), chrlabs = chr_labels)
                        abline(h = -log10(4.066333e-05), col = "red", lty = 2, lwd = 5)
dev.off()


png("qqplot_AMT2.jpg", width = 10000, height = 6000, units = "px", res=300)
par(mfrow=c(1,1), mar=c(8,8,5,5))
qq(pval.env3.ordered.by.amt$pval_AMT, main = "Q-Q plot - AMT",
   cex.lab=3, cex.axis=2, cex.main=3, cex.sub=3,
   col = c("blue4"))
   abline(h = -log10(4.066333e-05), col = "red", lty = 2, lwd = 5)
dev.off()

##############################################################################################################



```



#### AMP

```{r GWAS with LEA: AMP, include=FALSE}

##############################################################################################################
## Annual Mean Temperature (AMT)

## FDR of 0.05
q = 0.05

## Order sig.snps.amt by pval_AMT
pval.env3.ordered.by.amp <- chr.pval.env3[order(chr.pval.env3$pval_AMP),]

## Number of SNPs
L.amp = length(pval.env3.ordered.by.amp$pval_AMP)

## Adding the FDR adjusted pvalue
pval.env3.ordered.by.amp$BHadj.amp <- q*(1:L.amp)/L.amp
head(pval.env3.ordered.by.amp)

str( pval.env3.ordered.by.amp)
pval.env3.ordered.by.amp$CHROMF <- as.numeric(as.factor(pval.env3.ordered.by.amp$CHROM))
pval.env3.ordered.by.amp$POS <- as.numeric(pval.env3.ordered.by.amp$POS)

## Indices for Candidates SNPs
w.amp = which(pval.env3.ordered.by.amp$pval_AMP < q*(1:L.amp)/L.amp)
length(w.amp)

## Get rows in pval.env3.ordered.by.amt that correspond to the candidate SNPs
candidates.amp <- pval.env3.ordered.by.amp[w.amp,]
View(candidates.amp)


########################
## Manhattan and QQPlots

library(qqman)

## AMP
tail(candidates.amp)

png("manhattan_AMP2.jpg", width = 12000, height = 6000, units = "px", res=300)
par(mfrow=c(1,1), mar=c(8,8,5,5))
manhattan(pval.env3.ordered.by.amp, chr="CHROMF", bp="POS", snp="ID", 
                        genomewideline = F, suggestiveline = F,
                        logp = T, p="pval_AMP", annotateTop = T, main = "Manhattan Plot - AMP", 
                        ylim=c(0,10), cex.lab=3, cex.axis=2, cex.main=3, cex.sub=3,
                        col = c("blue4", "goldenrod1"), chrlabs = chr_labels)
                        abline(h = -log10(0.0002922521), col = "red", lty = 2, lwd = 5)
dev.off()


png("qqplot_AMP2.jpg", width = 10000, height = 6000, units = "px", res=300)
par(mfrow=c(1,1), mar=c(8,8,5,5))
qq(pval.env3.ordered.by.amp$pval_AMP, main = "Q-Q plot - AMP",
   cex.lab=3, cex.axis=2, cex.main=3, cex.sub=3,
   col = c("blue4"))
   abline(h = -log10(0.0002922521), col = "red", lty = 2, lwd = 5)
dev.off()

##############################################################################################################


```



#### PS

```{r GWAS with LEA: PS, include=FALSE}

##############################################################################################################
## Annual Mean Temperature (AMT)

## FDR of 0.05
q = 0.05

## Order sig.snps.amt by pval_AMT
pval.env3.ordered.by.ps <- chr.pval.env3[order(chr.pval.env3$pval_PS),]

## Number of SNPs
L.ps = length(pval.env3.ordered.by.ps$pval_PS)

## Adding the FDR adjusted pvalue
pval.env3.ordered.by.ps$BHadj.ps <- q*(1:L.ps)/L.ps
head(pval.env3.ordered.by.ps)

str( pval.env3.ordered.by.ps)
pval.env3.ordered.by.ps$CHROMF <- as.numeric(as.factor(pval.env3.ordered.by.ps$CHROM))
pval.env3.ordered.by.ps$POS <- as.numeric(pval.env3.ordered.by.ps$POS)

## Indices for Candidates SNPs
w.ps = which(pval.env3.ordered.by.ps$pval_PS < q*(1:L.ps)/L.ps)
length(w.ps)

## Get rows in pval.env3.ordered.by.amt that correspond to the candidate SNPs
candidates.ps <- pval.env3.ordered.by.ps[w.ps,]
View(candidates.ps)


########################
## Manhattan and QQPlots

library(qqman)

## AMP
tail(candidates.ps)

png("manhattan_PS2.jpg", width = 12000, height = 6000, units = "px", res=300)
par(mfrow=c(1,1), mar=c(8,8,5,5))
manhattan(pval.env3.ordered.by.ps, chr="CHROMF", bp="POS", snp="ID", 
                        genomewideline = F, suggestiveline = F,
                        logp = T, p="pval_PS", annotateTop = T, main = "Manhattan Plot - PS", 
                        ylim=c(0,10), cex.lab=3, cex.axis=2, cex.main=3, cex.sub=3,
                        col = c("blue4", "goldenrod1"), chrlabs = chr_labels)
                        abline(h = -log10(0.0006566112), col = "red", lty = 2, lwd = 5)
dev.off()


png("qqplot_PS2.jpg", width = 10000, height = 6000, units = "px", res=300)
par(mfrow=c(1,1), mar=c(8,8,5,5))
qq(pval.env3.ordered.by.ps$pval_PS, main = "Q-Q plot - PS",
   cex.lab=3, cex.axis=2, cex.main=3, cex.sub=3,
   col = c("blue4"))
   abline(h = -log10(0.0006566112), col = "red", lty = 2, lwd = 5)
dev.off()

```



```{r Some Descriptive Analysis, include=FALSE}
##############################################################################################################

## Summarizing number of rows by chromosome
library(dplyr)
library(reshape2)

## Adding column with name of cliamte variables
candidates.amt$var <- "AMT"
candidates.amp$var <- "AMP"
candidates.ps$var <- "PS"

## Changing last column name to pval
colnames(candidates.amt)[colnames(candidates.amt)=="BHadj.amt"] <- "pval"
colnames(candidates.amp)[colnames(candidates.amp)=="BHadj.amp"] <- "pval"
colnames(candidates.ps)[colnames(candidates.ps)=="BHadj.ps"] <- "pval"

## Merging all candidates
candidates.clim.go <- rbind(candidates.amt, candidates.amp, candidates.ps)
View(candidates.clim.go)
colnames(candidates.clim.go)

## Get the ID that are in the three variables
snp.3var <- candidates.clim.go %>% group_by(CHROM, ID) %>% filter(n() == 3)
dim(snp.3var)[1]/3
head(snp.3var)

## Get the number of rows per chromosome for  SNPs in two variables
snp.2var <- candidates.clim.go %>% group_by(CHROM, ID) %>% filter(n() == 2)
View(snp.2var)

## Get snp.2var in wide format (one row per SNP) var as columns and NA in the row for a variable that is not present
snp.2var.wide <- dcast(snp.2var, CHROM + ID ~ var, value.var = "pval")
View(snp.2var.wide)

## Add column with SNPs present in AMT and AMP
snp.2var.wide$AMT.AMP <- ifelse(is.na(snp.2var.wide$AMT), NA, ifelse(is.na(snp.2var.wide$AMP), NA, "AMT.AMP"))
snp.2var.wide$AMT.PS <- ifelse(is.na(snp.2var.wide$AMT), NA, ifelse(is.na(snp.2var.wide$PS), NA, "AMT.PS"))
snp.2var.wide$AMP.PS <- ifelse(is.na(snp.2var.wide$AMP), NA, ifelse(is.na(snp.2var.wide$PS), NA, "AMP.PS"))
View(snp.2var.wide)

## Get the number of SNPs per chromosome for AMT.AMP  that are not NA
snps.amt.amp <- snp.2var.wide %>% group_by(CHROM, AMT.AMP) %>% filter(!is.na(AMT.AMP)) %>% summarise(n=n())
View(snps.amt.amp)
sum(snps.amt.amp$n)

## Get the number of SNPs per chromosome for AMT.PS  that are not NA
snps.amt.ps <- snp.2var.wide %>% group_by(CHROM, AMT.PS) %>% filter(!is.na(AMT.PS)) %>% summarise(n=n())
View(snps.amt.ps)
sum(snps.amt.ps$n)

## Get the number of SNPs per chromosome for AMP.PS  that are not NA
snps.amp.ps <- snp.2var.wide %>% group_by(CHROM, AMP.PS) %>% filter(!is.na(AMP.PS)) %>% summarise(n=n())
View(snps.amp.ps)
sum(snps.amp.ps$n)

## Write table with candidates as csv file
write.csv(candidates.clim.go, file="candidates.csv", row.names = F)

## Get the number of SNPs per chromosome and var as columns
candidates.per.chr <- dcast(candidates.clim.go, CHROM + ID ~ var, value.var = "pval")
View(candidates.per.chr)

## Get total number of SNPs per chromosome
snps.per.chr <- candidates.per.chr %>% group_by(CHROM) %>% summarise(n=n())
View(snps.per.chr)

##### Number of SNPs per chromosome and variable
## See what chromosoe has more SNPs per variable

## Get the number of SNPs per chromosome and var as columns that are not NA
snps.chr.amt <- candidates.per.chr %>%
group_by(CHROM) %>% filter(!is.na(AMT)) %>% summarise(n=n())
View(snps.chr.amt)
sum(snps.chr.amt$n) ## check if the sum of the number of SNPs per chromosome is equal to the number of SNPs in AMT

snps.chr.amp <- candidates.per.chr %>%
group_by(CHROM) %>% filter(!is.na(AMP)) %>% summarise(n=n())
View(snps.chr.amp)
sum(snps.chr.amp$n) ## check if the sum of the number of SNPs per chromosome is equal to the number of SNPs in AMP

snps.chr.ps <- candidates.per.chr %>%
group_by(CHROM) %>% filter(!is.na(PS)) %>% summarise(n=n())
View(snps.chr.ps)
sum(snps.chr.ps$n) ## check if the sum of the number of SNPs per chromosome is equal to the number of SNPs in PS

## Get the number of rows per chromosome and variable var
candidates.per.chr.var <- candidates.clim.go %>% group_by(CHROM, var) %>% summarise(n=n())
View(candidates.per.chr.var)

## Write table with number of SNPs per chromosome as csv file
write.csv(candidates.per.chr.var, file="candidates_per_chr.csv", row.names = F)

## Making stylish barplot from candidates.per.chr.var
library(ggplot2)

chr_labels

## Making stylish barplot with number of SNPs per chromosome from candidates.per.chr.var with choromosome labels order by chr_labels
png("candidates_per_chr_var2.jpg", width = 12000, height = 6000, units = "px", res=300)
par(mfrow=c(1,1), mar=c(8,8,5,5))
ggplot(candidates.per.chr.var, aes(x=CHROM, y=n, fill=var)) + 
  geom_bar(stat="identity", position="dodge") +
  scale_fill_manual(values=c("blue4", "goldenrod1", "red")) +
  theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust=0.5, size=20)) +
  scale_x_discrete(limits=chr_labels) +
  theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust=0.5, size=20)) +
  theme(axis.text.y = element_text(size=20)) +
  theme(axis.title.x = element_text(size=20)) +
  theme(axis.title.y = element_text(size=20)) +
  theme(legend.title = element_text(size=20)) +
  theme(legend.text = element_text(size=20)) +
  theme(legend.position = "top") +
  theme(plot.title = element_text(size=20)) +
  theme(plot.subtitle = element_text(size=20)) +
  theme(plot.caption = element_text(size=20)) +
  labs(title="Number of Candidate SNPs per Chromosome and Climate Variable", 
       subtitle="FDR = 0.05") +
  theme(plot.title = element_text(hjust = 0.5)) +
  theme(plot.subtitle = element_text(hjust = 0.5)) +
  theme(plot.caption = element_text(hjust = 0.5)) +
  xlab("Chromosomes") + ylab("Number of SNPs")
dev.off()


```



```{r GENE ONTOLOGY, include=FALSE}
##############################################################################################################

## Read sequence_report.txt file with the length of all chromosomes of Taeniopygia guttata (zebra finch)
## This was downloaded from NCBI (https://www.ncbi.nlm.nih.gov/genome/?term=txid59729[orgn])
sequence.report <- read.table("sequence_report.txt", header = T, sep = "\t", stringsAsFactors = F)
View(sequence.report)
class(sequence.report)

## Remove rows with unknown chromosome
seq.report <- sequence.report[sequence.report$Chromosome.name != "Un",]

## Remove rows with empty "" chromosome
seq.report <- seq.report[seq.report$Chromosome.name != "",]

View(seq.report)

## Make a copy of candidates.clim.go
candidates.clim.go <- candidates.clim

## Cehck dataframe with candidates SNPs
View(candidates.clim.go)

## Replace X in CHROM by W
candidates.clim.go$CHROM[candidates.clim.go$CHROM == "X"] <- "W"
 
## ADD column in candidates.clim.go with seq.length ( length of the chromosome)
candidates.clim.go$seq.length <- seq.report$Seq.length[match(candidates.clim.go$CHROM, seq.report$Chromosome.name)]  

## Remove rows in candidates.clim.go with NA or empty values in CHROM and POS columns
candidates.clim.go <- candidates.clim.go[candidates.clim.go$CHROM != "NA" & candidates.clim.go$CHROM != "",]
dim(candidates.clim.go)

## Substract 1x10e6 to the POS column in a new column but the new value cannot be less than 0
candidates.clim.go$start <- candidates.clim.go$POS - 1e6
candidates.clim.go$start[candidates.clim.go$start < 0] <- 0

## Add 1x10e6 to the POS column in a new column if end is greater than seq.length, then end will be only seq.length
candidates.clim.go$end <- candidates.clim.go$POS + 1e6

## If end is greater than seq.length, then end will be only seq.length
candidates.clim.go$end <- ifelse(candidates.clim.go$end > candidates.clim.go$seq.length, candidates.clim.go$seq.length, candidates.clim.go$end)

## Make a new column pasting CHROM, start and end columns separated by semicolon
candidates.clim.go$chr.reg <- paste(candidates.clim.go$CHROM, candidates.clim.go$start, candidates.clim.go$end, sep=":")

## Extract only the chromosomal regions
chr.reg <- candidates.clim.go$chr.reg
View(candidates.clim.go)

## Group number of genes by var
colnames(candidates.clim.go)
var.gen <- candidates.clim.go %>% group_by(var) %>% summarise(n = n_distinct(external_gene_name))
View(var.gen)
## Get rows in candidates.clim.go of only AMT
candidates.clim.go.amt <- candidates.clim.go[candidates.clim.go$var == "AMT",]
dim(candidates.clim.go.amt)

## Get rows in candidates.clim.go of only AMP
candidates.clim.go.amp <- candidates.clim.go[candidates.clim.go$var == "AMP",]
dim(candidates.clim.go.amp)

## Get rows in candidates.clim.go of only PS
candidates.clim.go.ps <- candidates.clim.go[candidates.clim.go$var == "PS",]
dim(candidates.clim.go.ps)

## Get rows in candidates.clim.go of only PS and AMP
candidates.clim.go.ps.amp <- candidates.clim.go[candidates.clim.go$var == "PS" | candidates.clim.go$var == "AMP",]
dim(candidates.clim.go.ps.amp)


unique(length(chr.reg))


##############################################################################################################

## remove package dbplyr
remove.packages("dbplyr")


## Installing Biomart
if (!require("BiocManager", quietly = TRUE))
    install.packages("BiocManager")
BiocManager::install("biomaRt")
BiocManager::install("dbplyr")

library(biomaRt)
library(dbplyr)

## Connect to the desired BioMart database. The biomart argument should be given a valid name
ensembl <- biomaRt::useEnsembl(biomart = "genes")

## Check the dataset names
datasets <- biomaRt::listDatasets(ensembl)
View(datasets)

## Select the dataset
mart <- biomaRt::useEnsembl(biomart = "ensembl",
                  dataset = "tguttata_gene_ensembl")

gen.id <- biomaRt::getBM(attributes = c("entrezgene_id", "ensembl_gene_id", 
                              "external_gene_name", 'go_id', 'hgnc_symbol', 
                              "chromosome_name", "start_position","end_position", 'chromosomal_region'),
                        filters = c('chromosomal_region'),
                        values = chr.reg,
                        mart = mart)

## Checking results
View(head(gen.id))
View(gen.id)
dim(gen.id)
colnames(gen.id)
length(unique(gen.id$external_gene_name))

## NUmber of genes in the 7087 chromosomal regions
length(unique(gen.id$hgnc_symbol))

## Number of GO terms in the 7087 chromosomal regions
length(unique(gen.id$go_id))

## Check if ensembl_gene_id has NA
sum(is.na(gen.id$ensembl_gene_id))

library(dplyr)

## Summarize number of unique genes per chromosome in gen.id
unique(gen.id$chromosome_name)
chr.go.count <- gen.id  %>% group_by(chromosome_name) %>% summarize(n = n_distinct(go_id))
View(chr.go.count)

## Summarize number of unique go_id per chromosome in gen.id
unique(gen.id$chromosome_name)
chr.gen.count <- gen.id  %>% group_by(chromosome_name) %>% summarize(n = n_distinct(external_gene_name))
View(chr.gen.count)

## Merge candidates.clim.go and gen.id by chr.reg
colnames(candidates.clim.go)
colnames(gen.id)
candidates.clim.go.gen.id <- merge(candidates.clim.go, gen.id, by = "chr.reg")

library(tidyr)

## Make gen.id in long format putting each gene in the column description that is separated by "/" in different rows
gen.id.long <- gen.id %>% separate_rows(external_gene_name, sep = "/")
dim(gen.id.long)
head(gen.id.long)




################################################################################

## Installing clusterProfiler
if (!require("BiocManager", quietly = TRUE))
    install.packages("BiocManager")

BiocManager::install("clusterProfiler")
BiocManager::install("AnnotationHub")
BiocManager::install("rtracklayer")


library(clusterProfiler)

## Select uniques genes in gen.id
gene.name <- unique(gen.id$hgnc_symbol)

## REmove row with empty "" in gene.name
gene.name <- gene.name[gene.name != ""]

## Check length of gene.name
length(gene.name)
unique(gen.id$chromosome_name)

## Get the Galllus gallus data base
BiocManager::install("org.Gg.eg.db")
library(org.Gg.eg.db)

## Enrichment Analysis
wrengo <- clusterProfiler::enrichGO(gene          = gene.name,
                                    keyType = "SYMBOL",
                                    OrgDb         = "org.Gg.eg.db",
                                    ont           = "ALL",
                                    pAdjustMethod = "bonferroni",
                                    pvalueCutoff  = 0.05,
                                    #qvalueCutoff  = 0.1,
                                    readable      = TRUE)


## Check results
head(wrengo)
class(wrengo)
View(wrengo)

## Converting to data.frame
wrengo.dt <- as.data.frame(wrengo)

## Checking reuslts in the dataframe
View(wrengo.dt)
colnames(wrengo.dt)
head(wrengo.dt)
length(unique(wrengo.dt$Description))
length(unique(wrengo.dt$ID))

## Filter wrengo.dt where Ontology is BP
wrengo.bp <- wrengo.dt[wrengo.dt$ONTOLOGY == "BP",]
View(wrengo.bp)
length(unique(wrengo.bp$Description))
length(unique(wrengo.bp$ID))

## Select row in wrengo.bp with the lowest p.adjust value
View(wrengo.bp[wrengo.bp$p.adjust == min(wrengo.bp$p.adjust),])

## Filter wrengo.dt where Ontology is MF
wrengo.mf <- wrengo.dt[wrengo.dt$ONTOLOGY == "MF",]
View(wrengo.mf)
length(unique(wrengo.mf$Description))
length(unique(wrengo.mf$ID))

## Select row in wrengo.mf with the lowest p.adjust value
View(wrengo.mf[wrengo.mf$p.adjust == min(wrengo.mf$p.adjust),])


## Filter wrengo.dt where Ontology is CC
wrengo.cc <- wrengo.dt[wrengo.dt$ONTOLOGY == "CC",]
View(wrengo.cc)
length(unique(wrengo.cc$Description))
length(unique(wrengo.cc$ID))

## Select row in wrengo.cc with the lowest p.adjust value
View(wrengo.cc[wrengo.cc$p.adjust == min(wrengo.cc$p.adjust),])


## Plotting the results
BiocManager::install("enrichplot")
library(enrichplot)

## dotplot
dotplot(wrengo, showCategory=20)










```




```{r Saving, include=FALSE}

save.image("WGS_LEA.RData")

```


